### ISSUE TITLE
build-app.sh can't build cortex-m

### ISSUE THREAD

**Issue by honyczek** [lupyuen/pinetime-rust-mynewt#16]
When build-app.sh tries to compile `cortex-m` it ends with error:
```
+ mynewt_build_app=nrf52_my_sensor
+ rust_build_target=thumbv7em-none-eabihf
+ launch_config=launch-nrf52-pi.json
+ set +x

----- Building Rust app and Mynewt OS for thumbv7em-none-eabihf / nrf52_my_sensor...

----- Build Rust app
+ cargo build --target thumbv7em-none-eabihf
   Compiling cortex-m v0.6.2
error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/asm.rs:11:24
   |
11 |         () => unsafe { asm!("bkpt" :::: "volatile") },
   |                        ----^^^^^^^^^^^^^^^^^^^^^^^^
   |                        |
   |                        help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/asm.rs:39:13
   |
39 |               asm!("1:
   |               ^---
   |               |
   |  _____________help: replace with: `llvm_asm!`
   | |
40 | |                   nop
41 | |                   subs $0, $$1
42 | |                   bne.n 1b"
...  |
45 | |                  :
46 | |                  : "volatile");
   | |_______________________________^
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/asm.rs:68:24
   |
68 |         () => unsafe { asm!("nop" :::: "volatile") },
   |                        ----^^^^^^^^^^^^^^^^^^^^^^^
   |                        |
   |                        help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/asm.rs:89:24
   |
89 |         () => unsafe { asm!("wfe" :::: "volatile") },
   |                        ----^^^^^^^^^^^^^^^^^^^^^^^
   |                        |
   |                        help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
   --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/asm.rs:110:24
    |
110 |         () => unsafe { asm!("wfi" :::: "volatile") },
    |                        ----^^^^^^^^^^^^^^^^^^^^^^^
    |                        |
    |                        help: replace with: `llvm_asm!`
    |
    = note: consider migrating to the new asm! syntax specified in RFC 2873
    = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
   --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/asm.rs:131:24
    |
131 |         () => unsafe { asm!("sev" :::: "volatile") },
    |                        ----^^^^^^^^^^^^^^^^^^^^^^^
    |                        |
    |                        help: replace with: `llvm_asm!`
    |
    = note: consider migrating to the new asm! syntax specified in RFC 2873
    = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
   --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/asm.rs:155:24
    |
155 |         () => unsafe { asm!("isb 0xF" ::: "memory" : "volatile") },
    |                        ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                        |
    |                        help: replace with: `llvm_asm!`
    |
    = note: consider migrating to the new asm! syntax specified in RFC 2873
    = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
   --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/asm.rs:183:24
    |
183 |         () => unsafe { asm!("dsb 0xF" ::: "memory" : "volatile") },
    |                        ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                        |
    |                        help: replace with: `llvm_asm!`
    |
    = note: consider migrating to the new asm! syntax specified in RFC 2873
    = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
   --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/asm.rs:209:24
    |
209 |         () => unsafe { asm!("dmb 0xF" ::: "memory" : "volatile") },
    |                        ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                        |
    |                        help: replace with: `llvm_asm!`
    |
    = note: consider migrating to the new asm! syntax specified in RFC 2873
    = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/interrupt.rs:13:13
   |
13 |             asm!("cpsid i" ::: "memory" : "volatile");
   |             ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |             |
   |             help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/interrupt.rs:40:15
   |
40 |         () => asm!("cpsie i" ::: "memory" : "volatile"),
   |               ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |               |
   |               help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/basepri.rs:11:17
   |
11 |                 asm!("mrs $0, BASEPRI" : "=r"(r) ::: "volatile");
   |                 ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                 |
   |                 help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/basepri.rs:40:19
   |
40 |             () => asm!("msr BASEPRI, $0" :: "r"(_basepri) : "memory" : "volatile"),
   |                   ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                   |
   |                   help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/basepri_max.rs:17:23
   |
17 |                 () => asm!("msr BASEPRI_MAX, $0" :: "r"(_basepri) : "memory" : "volatile"),
   |                       ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                       |
   |                       help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
   --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/control.rs:166:30
    |
166 |                     unsafe { asm!("mrs $0, CONTROL" : "=r"(r) ::: "volatile") }
    |                              ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                              |
    |                              help: replace with: `llvm_asm!`
    |
    = note: consider migrating to the new asm! syntax specified in RFC 2873
    = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
   --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/control.rs:197:17
    |
197 |                 asm!("msr CONTROL, $0" :: "r"(control) : "memory" : "volatile");
    |                 ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                 |
    |                 help: replace with: `llvm_asm!`
    |
    = note: consider migrating to the new asm! syntax specified in RFC 2873
    = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/faultmask.rs:37:30
   |
37 |                     unsafe { asm!("mrs $0, FAULTMASK" : "=r"(r) ::: "volatile") }
   |                              ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                              |
   |                              help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/msp.rs:10:22
   |
10 |             unsafe { asm!("mrs $0,MSP" : "=r"(r) ::: "volatile") }
   |                      ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                      |
   |                      help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/msp.rs:33:15
   |
33 |         () => asm!("msr MSP,$0" :: "r"(_bits) :: "volatile"),
   |               ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |               |
   |               help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/primask.rs:37:30
   |
37 |                     unsafe { asm!("mrs $0, PRIMASK" : "=r"(r) ::: "volatile") }
   |                              ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                              |
   |                              help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/psp.rs:10:22
   |
10 |             unsafe { asm!("mrs $0,PSP" : "=r"(r) ::: "volatile") }
   |                      ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                      |
   |                      help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/psp.rs:33:15
   |
33 |         () => asm!("msr PSP,$0" :: "r"(_bits) :: "volatile"),
   |               ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |               |
   |               help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/apsr.rs:58:17
   |
58 |                 asm!("mrs $0, APSR" : "=r"(r) ::: "volatile");
   |                 ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                 |
   |                 help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/lr.rs:12:22
   |
12 |             unsafe { asm!("mov $0,R14" : "=r"(r) ::: "volatile") }
   |                      ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                      |
   |                      help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/lr.rs:28:15
   |
28 |         () => asm!("mov R14,$0" :: "r"(_bits) :: "volatile"),
   |               ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |               |
   |               help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/pc.rs:12:22
   |
12 |             unsafe { asm!("mov $0,R15" : "=r"(r) ::: "volatile") }
   |                      ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                      |
   |                      help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: the legacy LLVM-style asm! syntax is no longer supported
  --> /home/pi/.cargo/registry/src/github.com-1285ae84e5963aae/cortex-m-0.6.2/src/register/pc.rs:28:15
   |
28 |         () => asm!("mov R15,$0" :: "r"(_bits) :: "volatile"),
   |               ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |               |
   |               help: replace with: `llvm_asm!`
   |
   = note: consider migrating to the new asm! syntax specified in RFC 2873
   = note: alternatively, switch to llvm_asm! to keep your code working as it is

error: aborting due to 27 previous errors

error: could not compile `cortex-m`.

To learn more, run the command again with --verbose.

```
Expected behavior: Script compiles cortex-m

**Comment by lupyuen** [comment]
Hi: Sorry about that, can you run these commands...

```
   # Latest nightly-2020-04-20 fails with asm error, so we use nightly-2020-02-16
   source $HOME/.cargo/env
   rustup default nightly-2020-02-16
   rustup update
   rustup target add thumbv7em-none-eabihf
```

Then run the Build App script?

**Comment by honyczek** [comment]
Great, it works now!

**Comment by lupyuen** [comment]
Great thanks! I have updated my docs

**Comment by honyczek** [comment]
I've started from the clean Raspberry Pi again. I did steps according to README.md: 1), 3), 2) (reverse order, because scripts/install-pi.sh have to be downloaded first to run it), 4), 5), 6). But the script build-app.sh failed again with the same issue #16. After running commands from step 1) (or comment above) script finished successfully. Is there something in the scripts what breaks the script environment?

**Comment by lupyuen** [comment]
`install-pi` will reset the Rust version to the latest version, which will break the build.

So `rustup default nightly-2020-02-16` needs to be run after `install-pi`

Does that help?

**Comment by lupyuen** [comment]
I'm very sorry about the mess... Things are in transition right now...

1. We are preloading PineTime at the factory with FreeRTOS...

   https://github.com/JF002/Pinetime

   Together with the MCUBoot Bootloader that I have adapted. So I'm focusing my work on the FreeRTOS firmware with the MCUBoot Bootloader.

1.  The new PineTimes will no longer have Flash ROM protection. So there will be no longer any need for me to maintain the Flash Unprotect scripts (yay!). The new PineTimes will support firmware update over Bluetooth.

1.  I'm also not sure whether to keep maintaining Mynewt on PineTime. Since we are preloading FreeRTOS on PineTime and we have so few firmware developers, I'm considering moving over to FreeRTOS and dropping Mynewt.

Would you consider moving to FreeRTOS? I have simple instructions here for installing FreeRTOS...

https://gist.github.com/lupyuen/fb4cc04f03219d913cba8ec371f8c644

If you wish to stay on Mynewt, perhaps you can chat with me online so that you can get a quicker response. Here's the PineTime Chatroom on Discord / Matrix / Telegram / IRC. I'm usually online Singapore daytime (GMT+8 hours)...

https://wiki.pine64.org/index.php/PineTime#Community

Sorry again and thanks for using my code! :-)

**Comment by honyczek** [comment]
Ok, it is well described cause of issue, thanks.

I've got PineTime with factory OS with protection (bought in February 2020).

I've decided to flash your Mynewt firmware, because you had written very promising article on medium.com about flashing firmware using Raspberry Pi. I think you have done very good work including the helpful documentation. It's pity you're leaving Mynewt. But if there will be good docs for flashing FreeRTOS using Raspberry Pi, I'll follow you.

Thanks!
